kafka/conn.py

```diff
@@ -514,52 +514,45 @@ def _try_authenticate_gssapi(self, future):
         ctx_CanonName = ctx_Name.canonicalize(gssapi.MechType.kerberos)
         log.debug('%s: canonical Servicename: %s', self, ctx_CanonName)
         ctx_Context = gssapi.SecurityContext(name=ctx_CanonName, usage='initiate')
-        # Exchange tokens until authentication either succeeds or fails:
+        log.debug("%s: initiator name: %s", self, ctx_Context.initiator_name)
+
+        # Exchange tokens until authentication either succeeds or fails
         received_token = None
         try:
             while not ctx_Context.complete:
-                # calculate the output token
-                try:
-                    output_token = ctx_Context.step(received_token)
-                except GSSError as e:
-                    log.exception("%s: Error invalid token received from server",  self)
-                    error = Errors.ConnectionError("%s: %s" % (self, e))
+                # calculate an output token from kafka token (or None if first iteration)
+                output_token = ctx_Context.step(received_token)
 
-                if not output_token:
-                    if ctx_Context.complete:
-                        log.debug("%s: Security Context complete ", self)
-                    log.debug("%s: Successful GSSAPI handshake for %s", self, ctx_Context.initiator_name)
-                    break
+                # pass output token to kafka
                 try:
                     self._sock.setblocking(True)
-                    # Send output token
                     msg = output_token
                     size = Int32.encode(len(msg))
                     self._sock.sendall(size + msg)
-
                     # The server will send a token back. Processing of this token either
                     # establishes a security context, or it needs further token exchange.
                     # The gssapi will be able to identify the needed next step.
                     # The connection is closed on failure.
                     response = self._sock.recv(2000)
                     self._sock.setblocking(False)
 
-                except (AssertionError, ConnectionError) as e:
+                except ConnectionError as e:
                     log.exception("%s: Error receiving reply from server",  self)
                     error = Errors.ConnectionError("%s: %s" % (self, e))
-                    future.failure(error)
                     self.close(error=error)
+                    return future.failure(error)
 
                 # pass the received token back to gssapi, strip the first 4 bytes
+                # dpkp note: what are the first four bytes here?
```

Comment:
Yea it seems like it's the size there:
```
Client flow:
If sasl.mechanism is not GSSAPI, send a Kafka handshake request packet with the mechanism name to the server. Otherwise go to Step 3.
Request Format: | Kafka RequestHeader | Kafka SaslHandshakeRequest |
Wait for response from the server. If the error code in the response is non-zero, indicating failure, report the error and fail authentication.
Perform SASL authentication with the configured client mechanism. SASL authentication packets do not contain a Kafka RequestHeader.
Client token Format: | Size (int32) | SASL client authentication token |
Server flow:
Wait for first authentication packet from client
If this packet is a not valid Kafka handshake request, go to Step 4 and process this packet as the first GSSAPI client token
If the client mechanism in the Kafka handshake request received in Step 2 is enabled in the broker, send a response with error code zero and start authentication using the specified mechanism. Otherwise, send an error response including the list of enabled mechanisms and fail authentication.
Response Format: | Kafka ResponseHeader | Kafka SaslHandshakeResponse |
Perform SASL authentication with the selected mechanism. If mechanism exchange was skipped, process the initial packet that was received from the client first. SASL authentication packets are expected without a Kafka RequestHeader until SASL authentication exchange completes. SASL server authentication packets are sent back without a Kafka response header.
Server token Format: | Size (int32) | SASL server authentication token |
```
Source https://cwiki.apache.org/confluence/display/KAFKA/KIP-43%3A+Kafka+SASL+enhancements